lib/scholar/metrics.ex

```diff
@@ -29,34 +26,60 @@ defmodule Scholar.Metrics do
 
   ## Examples
 
-      iex> Scholar.Metrics.accuracy(Nx.tensor([[0, 1], [1, 0], [1, 0]]), Nx.tensor([[0, 1], [1, 0], [0, 1]]))
+      iex> Scholar.Metrics.binary_accuracy(Nx.tensor([1, 0, 0]), Nx.tensor([1, 0, 1]))
       #Nx.Tensor<
         f32
         0.6666666865348816
       >
 
   """
-  defn accuracy(y_true, y_pred) do
+  defn binary_accuracy(y_true, y_pred) do
     assert_shape(y_true, Nx.shape(y_pred))
 
-    transform({y_true, y_pred}, fn {y_true, y_pred} ->
-      if elem(Nx.shape(y_pred), Nx.rank(y_pred) - 1) == 1 do
-        y_pred
-        |> Nx.greater(0.5)
-        |> Nx.equal(y_true)
-        |> Nx.mean()
-      else
-        y_true
-        |> Nx.argmax(axis: -1)
-        |> Nx.equal(Nx.argmax(y_pred, axis: -1))
-        |> Nx.mean()
-      end
-    end)
+    y_pred
+    |> Nx.greater(0.5)
+    |> Nx.equal(y_true)
+    |> Nx.mean()
   end
 
   @doc ~S"""
-  Computes the precision of the given predictions with
-  respect to the given targets.
+  Computes the accuracy of the given predictions
+  for multi-class classification problems.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.accuracy(y_true, y_pred, num_classes: 3)
+      #Nx.Tensor<
+        f32
+        0.6000000238418579
+      >
+
+  """
+  defn accuracy(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes])
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    cm = confusion_matrix(y_true, y_pred, opts)
+
+    true_positive = Nx.take_diagonal(cm)
+    false_positive = Nx.sum(Nx.subtract(Nx.sum(cm, axes: [0]), true_positive))
+    false_negative = Nx.sum(Nx.subtract(Nx.sum(cm, axes: [1]), true_positive))
+    true_positive = Nx.sum(true_positive)
+
+    Nx.divide(true_positive, true_positive + 0.5 * (false_positive + false_negative) + 1.0e-16)
+  end
```

Comment:
Can't we do pretty much the same as the binary implementation?

```elixir
y_pred
|> Nx.equal(y_true)
|> Nx.mean()
```

Reply:
Also, the only difference is the thresholding we do in binary accuracy, but as I mentioned in the other comment, I'm not convinced we should do that. If we work with class numbers only, then both implementations are the same.

Reply:
you're right, the results are the same. Shall we expect the user pass normalised `y_pred` for all functions here? I think it's pretty easy to do `thresholded_preds` or convert one-hot encoded `y_pred`. They can be done using one single existing Nx APIs, i.e., `Nx.greater` or `Nx.argmax`. We probably don't need to have separate functions for them?

Reply:
Exactly, I think expecting normalized classes makes sense! We can totally have functions in some other module for converting back and forth to make the user code more readable, I just don't think this should be a concern of the metric calculation.

cc @seanmor5 for an opinion :)

Reply:
I created #13 for separating thresholding, we can handle it separately. We can still merge accuracy into a single function, and skip the `0.5` thresholding for binary :)

Reply:
I pushed a new commit that all functions expect normalized inputs. It was done last night but was waiting to see some other options. I agree a preprocess module could be used to do all the normalizing things instead of doing them in the metrics module.